### DISCUSSION TITLE
Docker Stack vs. Docker Service

### DISCUSSION THREAD

**Discussion by evanlalo** [BretFisher/ama#214] [Q&A] [answered]
First off, great content here and on your YouTube page, thanks for sharing!

## Overview
I'm rebuilding a core application for my business. It consists of three main containers running on a single host swarm.

1. Nginx container for reverse proxy and serves the frontend code
2. Python container for Django backend
3. MariaDB container

As of now, my main concerns are

1. Ability to rollback
2. Zero downtime

From what I can tell, both `stack` and `service` offer these and `stack` has been a whole lot easier to work with so I would prefer to use that.

## Question

Is there a big difference when running and updating these if I use `docker stack deploy` vs. `docker service create / update`?

Thanks!

**Comment by BretFisher** [comment, marked as answer]
No, they are identical. Technically in the background, the `docker stack` commands are turned into `service/volume/network` commands at the CLI and sent to the API as such. :)

**Reply by evanlalo** [reply]
Well.... that was quick and painless. Thanks for the info 👍

**Comment by chrisbecke** [comment]
It's worth mentioning that docker stack does not roll the entire stack back when a single service fails deployment. This can leave your stack in an inconsistent state and if you don't build your own alerting for it, you won't be told. There is also no `docker stack rollback` command to revert a failed deployment.
Also, stack deploy has no support for green/blue deployments, so during a stack deploy your v2 components will be communicating with v1 components.
